fix: decorator lag problems

SeriousBug opened the pull request:
Moves the computation of decorations and diagnostics to the engine. This significantly improves editing experience in long notes by eliminating the lag that happens if a long note is edited, then a command is used.

Docs (added config option to disable decorations): https://github.com/dendronhq/dendron-site/pull/311

Dev docs: https://github.com/dendronhq/dendron-docs/pull/2

kevinslin reviewed (commented):
traced through the calls and came up with the following:
- NOTE: some down my notes here -> [[Scratch|dendron://private/user.kevin.journal.2021.12.01.pr.kaan-decorator-lag.scratch]]

1. on an uncaught error:
- stub nodes do not have the `custom` prop, results in error when [isTaskNote](https://github.com/dendronhq/dendron/blob/e785efa8e2ce55bc39fb90cf34984d55035dd6ca/packages/common-all/src/types/configs/workspace/task.ts#L68) is called
    - we should have a `try/catch` around this block
- next task
    - we should modify stub notes to have a `custom` prop (I think this breaks our code in silent ways in other places as well)

2. On lag after input:
- it looks like its because of `findLinks`. trace of calls
    - WorkspaceWatcher.onDidChangeTextDocument
        - NoteSyncService.onDidChange
            - ...
                - findLinks
- see trace below when reunning against `dendron.tasks.backlog.md`:
    ```jsonc
    {"level":20,"time":1638405233007,"pid":13634,"hostname":"Kevins-Mac-mini.local","name":"dendron","ctx":"WorkspaceWatcher:onDidChangeTextDocument","uri":"/Users/kevinlin/Dropbox/Apps/Noah/workspaces/org-workspace-v2/org-private/dendron.tasks.backlog.md","state":"exit"}
    {"level":20,"time":1638405233007,"pid":13634,"hostname":"Kevins-Mac-mini.local","name":"dendron","ctx":"NoteSyncService:onDidChange","uri":"/Users/kevinlin/Dropbox/Apps/Noah/workspaces/org-workspace-v2/org-private/dendron.tasks.backlog.md"}
    {"level":30,"time":1638405233375,"pid":13634,"hostname":"Kevins-Mac-mini.local","name":"LinkUtils.getLinks","ctx":"getLinks","dlinksLength":6304,"noteRefsLength":16,"wikiLinksLength":6288}

    // 5 seconds btw `getLinks` and the call completing
    {"level":20,"time":1638405237425,"pid":13634,"hostname":"Kevins-Mac-mini.local","name":"dendron","ctx":"NoteSyncService:updateNoteContents","fname":"dendron.tasks.backlog","msg":"exit"}
    ```
- proper way to fix this is to do link calculation on the server side as well
    - I would do that in a separate commit though
    - meanwhile, an 80/20 might be to give up after say 2s of not finding a link 
        - mark this note as "too big" so we don't try to find the links 
        - add an info message (too many links in this document, some features are disabled, fix coming soon)
- future optimizations to consider
    - only calculate links on what has changed in the doc instead of the whole doc

SeriousBug commented on packages/common-all/package.json:
```diff
@@ -67,6 +67,7 @@
   "devDependencies": {
     "@types/js-yaml": "^3.12.5",
     "@types/nanoid-dictionary": "^4.2.0",
+    "@types/vscode": "^1.62.0",
```
I ended up not actually using the dependency, so I just removed it instead. Thanks for catching it!

SeriousBug commented:
I put the tasks for stub note `custom` prop and moving `findLinks` to engine to my backlog.

For `findLinks`, I just added the `maxNoteLength` check for now to fail it early for long notes. I didn't add a warning for this because Dendron already shows a warning for it during initialization.

Fixed the issue with the `isTaskNote` check, and also switched from throttling the `updateDecorations` to debouncing it. This was needed because with throttle the last update is missed, and the user is always left with stale decorations.

SeriousBug commented:
Finally, added some code to skip updating the decorations if the decorations we got from the engine are stale. Otherwise, placing these stale decorations causes decorations to appear in wrong spots in the document. If we instead don't apply these stale decorations, VSCode automatically shifts the existing decorations around in the document to update their locations. The downside is that the decorations never update while the user keeps typing, but I'm not sure how to handle this any better.

To see what I mean by this, remove the following portion from `windowDecorations.ts`:
```ts
    if (
      debouncedUpdateDecorations.states.get(
        updateDecorationsKeyFunction(editor)
      ) === "trailing"
    ) {
      // There's another execution that has already been called after this was
      // run. That means these results are stale. If existing lines have shifted
      // up or down since this function execution was started, setting the
      // decorations now will place the decorations at bad positions in the
      // document. On the other hand, if we do nothing VSCode will smartly move
      // those decorations to their new locations. With another execution
      // already scheduled, it's better to just wait for those decorations to
      // come in.
      return {};
    }
```

Then, open a long note like the backlog, select a position in the middle of the document, and start hitting enter. The decorations will start appearing in the middle of the document where the new lines have been entered.

kevinslin commented:
> 

this seems reasonable. when you say " decorations never update while the user keeps typing", given our current decorators, it doesn't seem like we would need any to update while the user types

- timestamp
- blockAnchor
- wikiLink
- brokenWikilink
- alias
- taskNote
